REGRESSION(272822@main): IPC sync replies might use freed data

kkinnunen-apple opened the pull request:
#### 83b7d1094b341c9fef824c58e88fcb5ebcf41141
<pre>
REGRESSION(272822@main): IPC sync replies might use freed data
<a href="https://bugs.webkit.org/show_bug.cgi?id=275434">https://bugs.webkit.org/show_bug.cgi?id=275434</a>
<a href="https://rdar.apple.com/125071066">rdar://125071066</a>

Reviewed by Cameron McCormack.

The commit 272822@main would remove IPC::Decoder from the
ConnectionSendSyncResult. The decoder owns the data pointed by the
data references. This change results in sender reading freed memory
when reading the reply data references.

Fix by ensuring ConnectionSendSyncResult stores the IPC::Decoder that
owns the data referenced by the reply.

Using data references, e.g. std::spans, in IPC messages means that the
data is stored in the IPC message. Contrast this with messages transferring
Vectors: the Vectors store the data.

* Source/WebKit/Platform/IPC/Connection.h:
(IPC::ConnectionSendSyncResult::ConnectionSendSyncResult):
(IPC::ConnectionSendSyncResult::reply):
(IPC::ConnectionSendSyncResult::takeReply):
(IPC::Connection::sendSync):
* Source/WebKit/Platform/IPC/MessageSender.h:
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::trySendSyncStream):
* Source/WebKit/UIProcess/RemotePageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::MockTestSyncMessageWithDataReply::name):
(TestWebKitAPI::MockTestSyncMessageWithDataReply::arguments):
(TestWebKitAPI::MockTestSyncMessageWithDataReply::MockTestSyncMessageWithDataReply):
(TestWebKitAPI::TEST_P):
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:

Canonical link: <a href="https://commits.webkit.org/280004@main">https://commits.webkit.org/280004@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5eb702f723056c6a5982f119da505cfbc144f30d

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/55490 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/34813 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/7956 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/58476 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/5922 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/57617 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/42434 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/6117 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/44685 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/4060 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/57518 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/32721 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/47821 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/25812 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/29506 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/5149 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/4066 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/51436 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/5417 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/60066 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/30644 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/5541 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/52118 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/31728 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/47891 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/51580 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/12288 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/32811 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/31477 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

heycam commented on Source/WebKit/Platform/IPC/Connection.h:
```diff
@@ -156,29 +156,33 @@ class WorkQueueMessageReceiver;
-    Expected<typename T::ReplyArguments, Error> value;
-
+// Sync message sender is expected to hold this instance alive as long as the reply() is being
+// accessed. View type data types in replies, such as std::span, refer to data stored in
```
s/View type data types/View data types/?

heycam commented on Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:
```diff
@@ -119,8 +119,17 @@ class MockConnectionClient final : public IPC::Connection::Client {
 
-    bool didReceiveSyncMessage(IPC::Connection&, IPC::Decoder&, UniqueRef<IPC::Encoder>&) override
+    // Handler contract as IPC::MessageReceiver::didReceiveSyncMessage: false on invalid message, may adopt encoder,
+    // decoder used only during the call, if encoder not adopted it will be submitted.
```
What does "submitted" mean here?

kkinnunen-apple replied:
Sending the encoder, e.g. sending a message

heycam reviewed (approved):
In the commit message:

s/would remove/removed/

kkinnunen-apple committed:
REGRESSION(272822@main): IPC sync replies might use freed data
https://bugs.webkit.org/show_bug.cgi?id=275434
rdar://125071066

Reviewed by Cameron McCormack.

The commit 272822@main would remove IPC::Decoder from the
ConnectionSendSyncResult. The decoder owns the data pointed by the
data references. This change results in sender reading freed memory
when reading the reply data references.

Fix by ensuring ConnectionSendSyncResult stores the IPC::Decoder that
owns the data referenced by the reply.

Using data references, e.g. std::spans, in IPC messages means that the
data is stored in the IPC message. Contrast this with messages transferring
Vectors: the Vectors store the data.

* Source/WebKit/Platform/IPC/Connection.h:
(IPC::ConnectionSendSyncResult::ConnectionSendSyncResult):
(IPC::ConnectionSendSyncResult::reply):
(IPC::ConnectionSendSyncResult::takeReply):
(IPC::Connection::sendSync):
* Source/WebKit/Platform/IPC/MessageSender.h:
* Source/WebKit/Platform/IPC/StreamClientConnection.h:
(IPC::StreamClientConnection::trySendSyncStream):
* Source/WebKit/UIProcess/RemotePageProxy.h:
* Source/WebKit/UIProcess/WebPageProxy.h:
* Tools/TestWebKitAPI/Tests/IPC/ConnectionTests.cpp:
(TestWebKitAPI::MockTestSyncMessageWithDataReply::name):
(TestWebKitAPI::MockTestSyncMessageWithDataReply::arguments):
(TestWebKitAPI::MockTestSyncMessageWithDataReply::MockTestSyncMessageWithDataReply):
(TestWebKitAPI::TEST_P):
* Tools/TestWebKitAPI/Tests/IPC/IPCTestUtilities.h:

Canonical link: https://commits.webkit.org/280004@main
